cas-compute/src/symbolic/derivative.rs

```diff
@@ -0,0 +1,268 @@
+
+fn product_rule(product: &[Expr], with: &str) -> Result<Expr, SymbolicDerivativeError> {
+    let mut outer_sum = SumBuilder::default();
+
+    // Produces a derivative according the product rule:
+    // f'*g*h + f*g'*h + f*g*h'
+    for derivative_index in 0..product.len() {
+        let mut inner_mult = MultBuilder::default();
+        for term_index in 0..product.len() {
+            let term = if derivative_index == term_index {
+                derivative(&product[derivative_index], with)?
+            } else {
+                product[term_index].clone()
+            };
+
+            inner_mult.mult(term);
+        }
+        
+        outer_sum.add(inner_mult.into());
+    }
+
+    Ok(outer_sum.into())
+}
+
+#[derive(Debug)]
+pub enum SymbolicDerivativeError {
+    /// The function may be differentiable, but we do not support symbolically computing it yet
+    Unsupported,
+
+    /// The provided sub-expression is not differentiable
+    Undifferentiable(Expr)
+}
+
+// Produces the derivative of the given expression, returning None if the derivative could not be
+// symbolically computed
+pub fn derivative(f: &Expr, with: &str) -> Result<Expr, SymbolicDerivativeError> {
+    if !non_zero(&f) {
+        return Ok(Expr::Primary(Primary::Integer(int(0))))
+    }
+    let expr = match f {
+        Expr::Primary(Primary::Float(_)) | Expr::Primary(Primary::Integer(_)) => {
+            Ok(Expr::Primary(Primary::Integer(int(0))))
+        },
+        Expr::Primary(Primary::Symbol(sym)) => {
+            if sym == with {
+                Ok(Expr::Primary(Primary::Integer(int(1))))
+            } else {
+                Ok(Expr::Primary(Primary::Integer(int(0))))
+            }
+        }
+        Expr::Primary(Primary::Call(func, args)) => {
+            // TODO: context and chain rule?
+            let mut mult_group = MultBuilder::default();
+            match func.as_str() {
+                "sqrt" => {
+                    assert_eq!(args.len(), 1, "sqrt has only one argument");
+                    return derivative(&Expr::Exp(Box::new(args[0].clone()), Box::new(Expr::Primary(Primary::Float(float(0.5))))), with);
+                },
+                "sin" => {
+                    assert_eq!(args.len(), 1, "sin has exactly one argument");
+                    mult_group.mult(derivative(&args[0], with)?);
+                    mult_group.mult(Expr::Primary(Primary::Call("sin".to_string(), vec![args[0].clone()])));
+                },
+                "cos" => {
+                    assert_eq!(args.len(), 1, "cos has exactly one argument");
+                    mult_group.mult(derivative(&args[0], with)?);
+                    mult_group.mult(Expr::Primary(Primary::Integer(int(-1))));
+                    mult_group.mult(Expr::Primary(Primary::Call("sin".to_string(), vec![args[0].clone()])));
+                },
+                _ => {
+                    return Err(SymbolicDerivativeError::Unsupported);
+                }
+            };
```

Comment:
Consider splitting this part of the implementation into a different file since there will be a lot of functions to work with in the future.